site-src/geps/gep-1058.md

```diff
@@ -0,0 +1,962 @@
+       `AllowedRoutes` criteria that the child route meets.
+5. A parent `Route` and any related child `Routes` must be the same type of
+   `Routes`.
+    1. Example: A child httpRoute can not point to a parent tcpRoute.
+6. A `Route` MUST NOT be a parent to a `Route` AND a child of another `Route`.
+    1. Explanation: This limitation avoids the need for specific
+       internal/terminal Route types, and the design of using a ParentRef from
+       the child to an AllowedRoutes slot on a parent allows a Route to know if
+       it is either a parent or child (or misconfigured as both) without
+       requiring the graph traversal that a design like ForwardTo may need to
+       avoid cycles.
+    2. If an implementation chooses to not enforce this limitation, that must be
+       documented along with how many tiers of parent/child route relationships
+       are allowed
+7. A Route’s `ParentRefs` can list more than one type of parent.
+    1. Example: A `HTTPRoute`’s ParentRefs can list one (or more) `Gateways` and
+       one (or more) `HTTPRoutes` at the same time.
+
+## Binding Parent and Child Routes
+
+Binding a child route to a parent route refers to the process of reconciling
+(i.e. combining) the two routes. Binding is treated as a single transaction for
+each Parent/Child route pair. Even when a child route attaches to multiple
+`RouteRules` in the parent route, the binding between the two routes is still a
+single transaction.
+
+The binding transaction has two stages; reconciliation and activation. The
+reconciliation process produces an Ephemeral route config. After the
+reconciliation process has completed successfully, the new Ephemeral config is
+made the active Ephemeral route config and applied to the data plane.
+
+An Ephemeral route is the result of the first child route to successfully bind
+to the parent route. After that, for each child route that attempts to bind to
+that parent route, the reconciliation process is actually between the existing
+Ephemeral route config and the child route's config. This is required because,
+evaluated individually, each parent/child reconciliation may produce a valid
+config but, all of them combined may produce an invalid config.
+
+This means that, when a parent route has multiple child routes that attach to
+it, the reconciliations of the children to the parent, must be done in series
+(as opposed to in parallel).
+
+NOTE: An Ephemeral route is not a formal Kubernetes object. It is a logical
+concept used, in this document, to help describe behavior. In some cases, the
+terms "parent" and "ephemeral" route are interchangeable depending on whether or
+not an ephemeral route already exists due to a previous, successful binding of a
+child to that parent.
+
+### Failed Binding
+
+The binding transaction MUST fail if there is A non-recoverable error during the
+binding process. A failed binding should result in no changes to the
```

Comment:
So say I have parent route v1 and child route v1. This is all valid config ("active Ephemeral").

Now, I change the child route to 'v2'. Its now invalid. Per the spec, I am supposed to use the old "active Ephemeral". 

Do I understand right? if so, this seems un-implementable without a stateful implementation storing prior state outside of k8s?

Reply:
I think what we've done elsewhere in the spec is probably relevant here - we should probably suggest that in the event of a conflict we detach/break immediately. We can still leave rooms for stateful implementations to provide a more graceful fallback if possible, but I don't think that's broadly implementable.

Reply:
There is another thread on 'roles' and permissions: I don't think a bad 'child' route in a delegated namespace should ever break or invalidate a 'parent' route, no matter when the parent is added. 

The first level of routes should be validated independently and be applied first, even if it conflicts with any child route. 

I would also require that child route validation and application is isolated at namespace boundary - I still don't know how delegation a /prefix to multiple namespaces works, but an error in one namespace should not impact a different one or prevent it from adding routes.